homeassistant/components/mysensors/gateway.py

```diff
@@ -58,48 +60,103 @@ def is_socket_address(value):
+    user_input_copy = user_input.copy()
+    try:
+        gateway: Optional[BaseAsyncGateway] = await _get_gateway(
+            hass, user_input_copy, str(randint(0, 10 ** 6)), persistence=False
+        )
+        if gateway is None:
+            return False
+        else:
+            gateway_ready = asyncio.Future()
+
+            def gateway_ready_callback(msg):
+                msg_type = msg.gateway.const.MessageType(msg.type)
+                _LOGGER.debug("Received MySensors msg type %s: %s", msg_type.name, msg)
+                if msg_type.name != "internal":
+                    return
+                internal = msg.gateway.const.Internal(msg.sub_type)
+                if internal.name != "I_GATEWAY_READY":
+                    return
+                _LOGGER.debug("Received gateway ready")
+                gateway_ready.set_result(True)
+
+            gateway.event_callback = gateway_ready_callback
+            connect_task = None
+            try:
+                connect_task = asyncio.create_task(gateway.start())
+                with async_timeout.timeout(5):
+                    await gateway_ready
+                    return True
+            except asyncio.TimeoutError:
+                _LOGGER.info("Try gateway connect failed with timeout")
+                return False
+            finally:
+                if connect_task is not None and not connect_task.done():
+                    connect_task.cancel()
+                asyncio.create_task(gateway.stop())
+    except OSError as err:
+        _LOGGER.info("Try gateway connect failed with exception", exc_info=err)
+        return False
+
+
+def get_mysensors_gateway(
+    hass: HomeAssistantType, gateway_id: GatewayId
+) -> Optional[BaseAsyncGateway]:
+    """Return the Gateway for a given GatewayId."""
+    if MYSENSORS_GATEWAYS not in hass.data[DOMAIN]:
+        hass.data[DOMAIN][MYSENSORS_GATEWAYS] = {}
+    gateways = hass.data[DOMAIN].get(MYSENSORS_GATEWAYS)
     return gateways.get(gateway_id)
 
 
-async def setup_gateways(hass, config):
+async def setup_gateway(
+    hass: HomeAssistantType, entry: ConfigEntry
+) -> Optional[BaseAsyncGateway]:
     """Set up all gateways."""
-    conf = config[DOMAIN]
-    gateways = {}
-
-    for index, gateway_conf in enumerate(conf[CONF_GATEWAYS]):
-        persistence_file = gateway_conf.get(
-            CONF_PERSISTENCE_FILE,
-            hass.config.path(f"mysensors{index + 1}.pickle"),
-        )
-        ready_gateway = await _get_gateway(hass, config, gateway_conf, persistence_file)
-        if ready_gateway is not None:
-            gateways[id(ready_gateway)] = ready_gateway
 
-    return gateways
+    ready_gateway = await _get_gateway(hass, entry)
+    return ready_gateway
 
 
-async def _get_gateway(hass, config, gateway_conf, persistence_file):
+async def _get_gateway(
+    hass: HomeAssistantType,
+    entry: Union[ConfigEntry, Dict[str, Any]],
```

Comment:
We should consolidate the signature to only accept a config entry and not a union.

Reply:
The dict variant is used by `try_connect()` which is executed before a config entry is ready.
How should i attempt to connect without the dict variant?

Reply:
Just pass the needed gateway parameters. It's better to decrease the scope of each function and not require the function to know about things not really related to its task.